File: mkdoc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#       mkdoc.py
#       
#       Copyright 2013 Átila Camurça <camurca.home@gmail.com>
#       
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#       
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#       
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.
#
import sys
import os
import shutil
import json

appdir, filename = os.path.split(os.path.abspath(__file__))
curdir = os.getcwd()
CONFIG_FILE = ".config.json"


def dir_is_empty():
    return os.listdir(curdir) == []


def init():
    sys.stdout.write("initializing project... ")
    if dir_is_empty():
        if len(sys.argv) > 2:
            type = sys.argv[2]
        else:
            type = "beamer"
        shutil.copy(appdir + "/" + type + ".tex", curdir + "/main.tex")
        filename = curdir + "/content.md"
        file = open(filename, 'w')
        file.write("""<!-- your text here -->""")
        file.close()
        # config file
        config = {"type": type}
        data_string = json.dumps(config)
        filename = "%s/%s" % (curdir, CONFIG_FILE)
        file = open(filename, 'w')
        file.write(data_string)
        file.close()
        print "done."
    else:
        print "[ERROR] directory is not empty."


def edit():
    print 'editing project ...'
    if len(sys.argv) > 2:
        opt = sys.argv[2]
        filename = 'main.tex'
        if not opt in ('-t', '--template'): print '[ERROR] option %s not found.' % opt; return
    else:
        filename = 'content.md'
    editor = raw_input('open with [blank for default]: ')
    if not editor:
        editor = 'xdg-open'
    os.system("%s %s &" % (editor, filename))


def view():
    json_data = open(CONFIG_FILE)
    data = json.load(json_data)
    print "compiling project ..."
    os.system("pandoc -t %s content.md -o content.tex" % data["type"])
    os.system("pdflatex -shell-escape -interaction=nonstopmode main.tex")
    os.system("xdg-open main.pdf")
    json_data.close()


def editor():
    print "opening mkdoc-editor ... "
    editor_path = "%s/editor/" % appdir
    os.system("xdg-open http://localhost:9669")
    os.system("cd %s && nodemon server.js %s" % (editor_path, curdir))


def help():
    print """
mkdoc(1)

USAGE
    mkdoc command [options]

COMMANDS
    init - start a project in an empty directory
        options: beamer|latex
        default: beamer

    editor - open a web editor

    edit - open file to edit
        options:
        -t, --template - open template file to edit instead

    view - generate file and open pdf

    help - show this help

    docs - show the documentation of mkdoc in the default browser

    cleanup - remove log and aux files
"""


def docs():
    url = "file://%s/docs/index.html" % appdir
    os.system("xdg-open %s" % url)


def cleanup():
    os.system("cd %s && ls *.aux *.log *.nav *.out *.snm *.toc" % curdir)
    answer = raw_input("Confirm remove? _yes/[no]: ")
    if answer in ('y', 'yes'):
        os.system("cd %s && rm *.aux *.log *.nav *.out *.snm *.toc" % curdir)

if __name__ == '__main__':
    opts = {'init': init, 'edit': edit, 'view': view, 'help': help,
            'docs': docs, 'cleanup': cleanup, 'editor': editor}
    if len(sys.argv) > 1:
        command = sys.argv[1]
    else:
        command = None

    opts.get(command, help)()

File: mkdoc.js
#!/usr/bin/env node

var program 	= require('commander'),
	 fs 		= require("fs"),
	 util 		= require("util"),
	 path		= require("path"),
	 exec		= require('child_process').exec,
	 prompt		= require("prompt"),
	 sh 		= require('execSync');

// TODO: make a file with all constants
var CONFIG_FILENAME = ".config.json",
	 SUPPORTED_TYPES = ['beamer', 'latex', 'io-slides'];

prompt.message = "mkdoc";
prompt.start();

var mkdoc = (function() {

	function init(type) {
		console.log("[ INFO] initializing project... ");
		type = type || "beamer";

		if (SUPPORTED_TYPES.indexOf(type) === -1) {
			console.log("[ERROR] Type %s not supported.", type);
			console.log("[ INFO] Type can be: %s", SUPPORTED_TYPES.join(', '));
			return;
		}

		var callback = _presentationType(type);
		callback.init();
	}

	function editor() {
		console.log("[ INFO] opening mkdoc-editor ... ");
		// FIXME: I change the order because the server block the xdg-open :(
		// TODO: make a callback function to call xdg-open before server starts!
		var cmd = util.format("xdg-open http://localhost:9669 && cd %s/editor && nodemon server.js %s", __dirname, process.cwd());
		sh.run(cmd);
	}

	function edit(file) {
		file = file || "content.md";
		console.log("[ INFO] editing '%s' ...", file);

		var extname = path.extname(file);
		if (extname !== ".md" && file !== "main.tex")	{
			console.log("[ INFO] you should open only markdown files (.md)");
		}

		fs.exists(file, function (exists) {
			if (exists) {
				prompt.get({
					properties: {
						app: {
							description: "open with [blank for default]"
						}
					}
				}, function(err, result) {
					result.app = result.app || "xdg-open";
					// TODO: look for another way to detach the process from the console
					// gedit detach himself :)
					var cmd = util.format("%s %s &", result.app, file);
					exec(cmd, function(err, stdout, stderr) {
						if (err) { throw err; }
					});
				});
			} else {
				console.log(util.format("[ERROR] file '%s' not found.", file));
			}
		});
	}

	function view() {
		console.log("[ INFO] compiling project ...");
		var callback = _presentationType();
		callback.view();
	}

	function docs() {
		console.log("[ INFO] opening docs ...");
		var url = util.format("file://%s/editor/docs.html", __dirname);
		sh.run(util.format("xdg-open %s", url));
	}

	function cleanup() {
		console.log("[ INFO] cleaning up ...");
        var callback = _presentationType();
		callback.cleanup();
	}

	function template() {
		edit("main.tex");
	}

	/* private functions */

	/**
	 * call require based on type, if type is undefined read the .config.json file.
	 * @param {[String]} type [presentation file (optional)]
	 */
	function _presentationType(type) {
		if (!type) {
			var config_filename = util.format("%s/%s", process.cwd(), CONFIG_FILENAME);
			var data = fs.readFileSync(config_filename);
			var config = JSON.parse(data);
			type = config.type;
		}
		return require("./" + type + ".js");
	}

	/* end of private functions */

	/* public functions */
	return {
		init: init,
		editor: editor,
		edit: edit,
		view: view,
		docs: docs,
		cleanup: cleanup,
		template: template,
	};
})();

program.version("1.0.0");
program.command("init [type]")
		.description("start a new project in an empty directory")
		.action(mkdoc.init);
program.command("editor")
		.description("open a web editor")
		.action(mkdoc.editor);
program.command("edit [file]")
		.description("open file to edit")
		.action(mkdoc.edit);
program.command("template")
		.description("open template file to edit")
		.action(mkdoc.template);
program.command("view")
		.description("compile files and open pdf")
		.action(mkdoc.view);
program.command("docs")
		.description("show the documentation of mkdoc in the browser")
		.action(mkdoc.docs);
program.command("cleanup")
		.description("remove log and aux files")
		.action(mkdoc.cleanup);

program.parse(process.argv);

if (!program.args.length) { program.help();}
